peps/pep-0747.rst

```diff
@@ -0,0 +1,903 @@
+variable within the same function definition:
+
+::
+
+   def as_type[T](typx: TypeExpr[T]) -> type[T] | None:
+       return typx if isinstance(typx, type) else None
+
+
+Interactions with TypeIs[] and TypeGuard[]
+------------------------------------------
+
+A type variable constrained by ``TypeExpr[]`` can also be used by a ``TypeIs[]``
+within the same function definition:
+
+::
+
+   def isassignable[T](value: object, typx: TypeExpr[T]) -> TypeIs[T]: ...
+
+   count: int | str = ...
+   if isassignable(count, int):
+       assert_type(count, int)
+   else:
+       assert_type(count, str)
+
+or by a ``TypeGuard[]`` within the same function definition:
+
+::
+
+   def isdefault[T](value: object, typx: TypeExpr[T]) -> TypeGuard[T]:
+       return (value == typx()) if isinstance(typx, type) else False
+
+   value: int | str = ''
+   if isdefault(value, int):
+       assert_type(value, int)
+       assert 0 == value
+   elif isdefault(value, str):
+       assert_type(value, str)
+       assert '' == value
+   else:
+       assert_type(value, int | str)
+
+
+.. _interactions_with_annotated:
+
+Interactions with Annotated[]
+-----------------------------
+
+``Annotated[]`` forms preserve their metadata at runtime:
+
+::
+
+   >>> ValueRange: TypeAlias = slice
+   >>> PositiveInt: TypeAlias = Annotated[int, ValueRange(1, float('inf'))]
+   >>> PositiveInt
+   typing.Annotated[int, slice(1, inf, None)]  # NOT: int
+
+However ``Annotated[]`` forms need not preserve their metadata at
+typechecking-time:
+
+::
+
+   count: PositiveInt = 1
+   assert_type(count, int)  # NOT: Annotated[int, ValueRange(1, float('inf'))]
+
+In particular when an ``Annotated[]`` argument is passed to a ``TypeExpr[]``
+parameter constraining a type variable that is also used by a ``TypeIs[]``
+or ``TypeGuard[]``, only the first type argument to ``Annotated[]`` must be
+maintained in the type inferred by a type checker. The second metadata
+argument to ``Annotated[]`` does need not be maintained:
+
+::
+
+   # Similar to isassignable(), but accepts Annotated[] forms describing constraints
+   def ismatch[T](value: object, typx: TypeExpr[T]) -> TypeGuard[T]: ...
+
+   count: int | str = -1
+   if ismatch(count, PositiveInt):
+       assert_type(count, int)  # NOT: Annotated[int, ValueRange(1, float('inf'))]
+   else:
+       assert_type(count, int | str)
+
+The above behavior of stripping the metadata argument to ``Annotated[]`` is
+intended to be consistent with the current behavior of static type
+checkers. Future PEPs may be interested in altering the behavior to
+preserve the metadata argument to ``Annotated[]``, but that is out of scope
+of this PEP.
+
+
+Interactions with isinstance() and issubclass()
+-----------------------------------------------
+
+The ``TypeExpr`` special form is not a ``type`` at runtime. It is
+however an ``object``:
+
+::
+
+   >>> isinstance(TypeExpr, type)
+   False
+
+   >>> isinstance(TypeExpr, object)
+   True
```

Comment:
I don't think we need to specify how TypeExpr is implemented at runtime. I would remove this sentence.